List `-debug` packages missing from AUR separately (Split Debug Package Support)

Ferdi265 opened the pull request:
This PR makes it so `-debug` packages created via the `(debug strip)` makepkg option combination are listed separately from regular packages in the "Missing AUR Packages" warning.

The rationale for this is, that if someone uses this option combination, every AUR package will have a `$pkgbase-debug` counterpart containing source files and debug symbols. These `$pkgbase-debug` packages obviously do not exist in the AUR, so are always listed in the "Missing AUR Packages" warning, making legitimately missing packages hard to spot.

Since there is no way (that I know of) to identify these debug packages other than by their name, this invariably will lead to false-positives (e.g. `python-flask-debug`, `dosbox-debug`, or `godot-debug`), which is why I opted to show them as a separate warning rather than dropping them entirely from the list.

As for the implementation, this small patch just separates the packages directly in the print function, which works well, but admittedly is a bit ugly.

I appreciate discussion on whether `yay` actually wants to handle this, and/or if `-debug` packages should get some way to identify them properly (this would require upstream changes to `makepkg` and possibly `pacman`). Split Debug Packages are currently not very well supported in general, and also not documented in many places. I have marked the PR as [RFC] until the design is clear.

(As an added note, yay grew support for installing these `-debug` packages in PR #1191, also by me)

Ferdi265 committed:
feat(query): list debug packages missing from AUR separately

Jguer commented on pkg/query/aur_warnings.go:
```diff
@@ -37,6 +45,21 @@ func (warnings *AURWarnings) Print() {
 	}
 }
 
+func filterDebugPkgs(names []string) (normal, debug []string) {
```
```suggestion
func filterDebugPkgs(names []string) ([]string, []string) {
```

Ferdi265 replied:
addressed in c23def7

Jguer commented on pkg/query/aur_warnings.go:
```diff
@@ -37,6 +45,21 @@ func (warnings *AURWarnings) Print() {
 
+func filterDebugPkgs(names []string) (normal, debug []string) {
+	normal = make([]string, 0)
+	debug = make([]string, 0)
```
```suggestion
	debug = make([]string, 0, len(names))
```

Ferdi265 replied:
addressed in c23def7

Jguer commented on pkg/query/aur_warnings.go:
```diff
@@ -37,6 +45,21 @@ func (warnings *AURWarnings) Print() {
 }
 
+func filterDebugPkgs(names []string) (normal, debug []string) {
+	normal = make([]string, 0)
```
```suggestion
	normal = make([]string, 0, len(names))
```

Ferdi265 replied:
addressed in c23def7

Jguer commented on pkg/query/aur_warnings.go:
```diff
@@ -37,6 +45,21 @@ func (warnings *AURWarnings) Print() {
+		}
+	}
+
+	return
```
```suggestion
	return normal, debug
```

Ferdi265 replied:
addressed in c23def7

Jguer reviewed (changes requested):
I know it's less clear for the returns like this but it avoids reallocating the slice at the beginning after it's already been declared and it allocates the max capacity of those slices

Jguer commented:
> I appreciate discussion on whether yay actually wants to handle this, and/or if -debug packages should get some way to identify them properly (this would require upstream changes to makepkg and possibly pacman). Split Debug Packages are currently not very well supported in general, and also not documented in many places. I have marked the PR as [RFC] until the design is clear.

This ambiguity in them is the reason why I feel hackish solutions are fine for them :sweat_smile:

Jguer commented on pkg/query/aur_warnings.go:
```diff
@@ -21,9 +22,16 @@ func NewWarnings() *AURWarnings {
+	}
+
+	if len(debugMissing) > 0 {
+		text.Warn(gotext.Get("Missing AUR Debug Packages:"))
```
You might have to regenerate the english .po file for this, if you need help with it just ping me and I'll add it after

Ferdi265 replied:
I have looked in the repo to see if there are simple instructions on how to regenerate the .po file, but didn't find a lot.
I'd appreciate some help with that, yes

Ferdi265 commented:
Thanks for the review, I'll address the changes to the filter function in a commit shortly

Ferdi265 committed:
fix(query): allocate capacity in filterDebugPkgs

(addresses review in PR #1484 by Jguer)

Ferdi265 commented:
The latest push should have addressed your change requests (minus the .po changes).
The lint check failed because of missing `diff` command in the container somehow, haven't seen that happen before.

Ferdi265 commented:
> This ambiguity in them is the reason why I feel hackish solutions are fine for them :sweat_smile:

That is a valid way to do it IMO. Better deal with the common cases in a nice way and make it possible to deal with the odd ones out, but don't pessimize the common case for them.

Removing the [RFC] label then :)

Jguer commented:
Broken lint stage is on me, I'll have it fixed

Ferdi265 commented:
sorry for the push clutter; I just wanted to check if the linting worked already.

Anyway, the branch is pretty much ready, except that I don't know how to regenerate the `.po` translation files (I presume there's a better way than to manually add the entry)

Jguer commented:
@Ferdi265 the CI should be working now as expected

Ferdi265 commented:
Rebased; the new linter seems to like the named return style more, so I reverted it back to that.

(Link to lint result: https://github.com/Jguer/yay/runs/2334122378)

Jguer commented:
Let's take it for a spin, thanks for the contribution @Ferdi265
